symengine/functions.cpp

```diff
@@ -3142,4 +3140,238 @@ RCP<const Basic> abs(const RCP<const Basic> &arg)
+        if (is_a<Complex>(*p) or is_a<Max>(*p))
+            return false;
+        if (not is_a_Number(*p))
+            non_number_exists = true;
+    }
+    if (not std::is_sorted(arg.begin(), arg.end(), RCPBasicKeyLess()))
+        return false;
+
+    return non_number_exists ;   // all arguments cant be numbers
+}
+
+bool Max::__eq__(const Basic &o) const
+{
+    if (is_a<Max>(o) and
+        vec_basic_eq_perm(arg_, static_cast<const Max &>(o).arg_))
+        return true;
+    else
+        return false;
+}
+
+int Max::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<Max>(o))
+    const Max &s = static_cast<const Max &>(o);
+    return vec_basic_compare(arg_, s.arg_);
+}
+
+std::size_t Max::__hash__() const
+{
+    std::size_t seed = MAX;
+    for (const auto &p: arg_) {
+        hash_combine<Basic>(seed, *p);
+    }
+    return seed;
+}
+
+RCP<const Basic> max(const vec_basic &arg)
+{
+    bool number_set = false;
+    RCP<const Number> max_number, difference;
+    set_basic new_args;
+
+    for (const auto &p: arg) {
+        if (is_a<Complex>(*p))
+            throw std::runtime_error("Complex can't be passed to max!");
+
+        if (is_a_Number(*p)) {
+            if(not number_set) {
+                max_number = rcp_static_cast<const Number>(p);
+
+            } else {
+                difference = rcp_static_cast<const Number>(p)->sub(*max_number);
+
+                if (difference->is_zero() and not difference->is_exact()) {
+                    if (max_number->is_exact())
+                        max_number = rcp_static_cast<const Number>(p);
+                } else if (difference->is_positive()) {
+                    max_number = rcp_static_cast<const Number>(p);
+                }
+            }
+            number_set = true;
+
+        } else if (is_a<Max>(*p)) {
+            for (const auto &l: rcp_static_cast<const Max>(p)->get_args()) {
+                if (is_a_Number(*l)) {
+                    if(not number_set) {
+                        max_number = rcp_static_cast<const Number>(l);
+
+                    } else {
+                        difference = rcp_static_cast<const Number>(l)->sub(*max_number);
+
+                        if (difference->is_zero() and not difference->is_exact()) {
+                            if (max_number->is_exact())
+                                max_number = rcp_static_cast<const Number>(l);
+                        } else if (difference->is_positive()) {
+                            max_number = rcp_static_cast<const Number>(l);
+                        }
+                    }
+                    number_set = true;
+                } else {
+                    new_args.insert(l);
+                }
+            }
+        } else {
+            new_args.insert(p);
+        }
+    }
+
+    if (number_set)
+        new_args.insert(max_number);
+
+    vec_basic final_args(new_args.size());
+    std::copy(new_args.begin(), new_args.end(), final_args.begin());
```

Comment:
Yes, it's a waste to do this copy. `vec_basic` I think might be easier to use, we can always construct `Max` directly if needed for speed (Debug time checked) so I think it's fine for now, we might want to revisit later.